runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java

```diff
@@ -829,6 +864,14 @@ public Instant get() {
   @GuardedBy("refreshLock")
   private final Set<ExecutableT> pendingRefreshes;
 
+  /**
+   * A set of executables with currently extracted timers, that are to be processed. Note that, due
```

Comment:
Consistency should only be per (key, window) pair.

Reply:
That is correct. The problem is that `WatermarkManager` works on level of `PTransform`s. I can push the logic down to `TransformWatermarks`, which is the first level, that is somewhat aware of keys (in `extractFiredTimers`), but that would mean, that timers extracted from `AppliedPTransformInputWatermark` would have to be setup again. That is because the logic requires that extracted timer to be fired, or setup again (in the way that `StatefulParDoEvaluator` pushes them back). I think that this makes the logic far too much complicated given that what we gain is the ability to extract and execute timers for various keys of the same PTransform in parallel. I must emphasize, that this is related only to DirectRunner, which is already pretty slow, because it does many validations and its main purpose is testing. And moreover, I'm not convinced that implementing this optimization in reality would bring and significant performance boost (speculative claim, I didn't do any performance testing). Maybe we can merge it as it is and solve performance issues after if they appear (avoid premature optimization with significant impact on readability and understandability of already somewhat complicated code)?